cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+            known_devices.populate_qubit_pairs_in_device_proto(pairs, out)
+        _serialize_gateset_and_gate_durations(
+            out, gateset, {} if gate_durations is None else gate_durations
+        )
+        _validate_device_specification(out)
+
+        return out
+
+    @classmethod
+    def from_device_information(
+        cls,
+        *,
+        qubit_pairs: Collection[Tuple[cirq.GridQubit, cirq.GridQubit]],
+        gateset: cirq.Gateset,
+        gate_durations: Optional[Mapping['cirq.GateFamily', 'cirq.Duration']] = None,
+    ) -> 'GridDevice':
+        """Constructs a GridDevice using the device information provided.
+
+        This is a convenience method for constructing a GridDevice given partial gateset and
+        gate_duration information: for every distinct gate, only one representation needs to be in
+        gateset and gate_duration. The remaining representations will be automatically generated.
+
+        For example, if the input gateset contains only `cirq.PhasedXZGate`, and the input
+        gate_durations is `{cirq.GateFamily(cirq.PhasedXZGate): cirq.Duration(picos=3)}`,
+        `GridDevice.metadata.gateset` will be
+
+        ```
+        cirq.Gateset(cirq.PhasedXZGate, cirq.XPowGate, cirq.YPowGate, cirq.PhasedXPowGate)
+        ```
+
+        and `GridDevice.metadata.gate_durations` will be
+
+        ```
+        {
+            cirq.GateFamily(cirq.PhasedXZGate): cirq.Duration(picos=3),
+            cirq.GateFamily(cirq.XPowGate): cirq.Duration(picos=3),
+            cirq.GateFamily(cirq.YPowGate): cirq.Duration(picos=3),
+            cirq.GateFamily(cirq.PhasedXPowGate): cirq.Duration(picos=3),
+        }
+        ```
+
+        This method reduces the complexity of constructing `GridDevice` on server side by requiring
+        only the bare essential device information.
+
+        Args:
+            qubit_pairs: Collection of bidirectional qubit couplings available on the device.
+            gateset: The gate set supported by the device.
+            gate_durations: Optional mapping from gates supported by the device to their timing
+                estimates. Not every gate is required to have an associated duration.
+            out: If set, device information will be serialized into this DeviceSpecification.
+
+        Raises:
+            ValueError: If a pair contains two identical qubits.
+            ValueError: If `gateset` contains invalid GridDevice gates.
+            ValueError: If `gate_durations` contains keys which are not in `gateset`.
+            ValueError: If multiple gate families in gate_durations can
+                represent a particular gate, but they have different durations.
+        """
+        metadata = cirq.GridDeviceMetadata(
+            qubit_pairs=qubit_pairs, gateset=gateset, gate_durations=gate_durations
+        )
+        incomplete_device = GridDevice(metadata)
+        return GridDevice.from_proto(incomplete_device.to_proto())
```

Comment:
This feels odd to create a device from metadata, then turn it into a proto and then back into a device. Is there a reason for the roundtrip through protos here? If so a comment would be helpful.

Reply:
Added - lmk if the rationale doesn't make sense